File: ios/Classes/CalendarEventsPlugin.swift
import Flutter
import UIKit

public class CalendarEventsPlugin: NSObject, FlutterPlugin {
    
    let calendarManager :CalendarEventManager = CalendarEventsManagerImpl()
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.godwin/calendar_events", binaryMessenger: registrar.messenger())
        let instance = CalendarEventsPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "checkPermission":
            let permission = calendarManager.checkPermission()
            result(permission)
            break
        case "requestPermission":
            calendarManager.requestPermission{intValue in  result(intValue)}
            break
        case "addEvent":
            if let map = call.arguments as? [String: Any] {
                let calendarEvent =  CalendarUtil.convertToEvent(from: map)
                let eventresult: CalendarEventResult = calendarManager.addEvent(calendarEvent: calendarEvent)
                switch eventresult{
                case .success(let event) :
                    result(["eventId": event.eventId])
                    break
                case .failed(let error) :
                    result(FlutterError(code:error.errorCode,message: error.details,details: nil))
                    break
                case .deleteSuccess(eventId: _):
                    break
                }
            } else {
                result(FlutterError(
                    code: "ARGUMENT_ERROR",
                    message: "Invalid arguments",
                    details: nil
                ))
            }
            break
        case "updateEvent":
            if let map = call.arguments as? [String: Any] {
                let calendarEvent =  CalendarUtil.convertToEvent(from: map)
                let eventresult = calendarManager.updateEvent(calendarEvent: calendarEvent)
                switch eventresult{
                case .success(let event) :
                    result(["eventId": event.eventId])
                    break
                case .failed(let error) :
                    result(FlutterError(code:error.errorCode,message: error.details,details: nil))
                    break
                case .deleteSuccess(eventId: _):
                    break
                }
                
            } else {
                result(FlutterError(
                    code: "ARGUMENT_ERROR",
                    message: "Invalid arguments",
                    details: nil
                ))
            }
            break
        case "deleteEvent":
            if let map = call.arguments as? [String: Any] {
                let eventId = map["eventId"] as? String
                let eventresult = calendarManager.deleteEvent(eventId: eventId)
                switch eventresult{
                case .success(_) :
                    break
                case .failed(let error) :
                    result(FlutterError(code:error.errorCode,message: error.details,details: nil))
                    break
                case .deleteSuccess(eventId: let eventId):
                    result(["eventId": eventId])
                }
            } else {
                result(FlutterError(
                    code: "ARGUMENT_ERROR",
                    message: "Invalid arguments",
                    details: nil
                ))
            }
            break
        case "getCalendarAccounts":
            let calendarResult  = calendarManager.getCalendarAccounts()
            switch calendarResult {
            case .success(let list):
                result(list)
            case .failed(let error):
                result(FlutterError(
                    code: error.errorCode,
                    message: error.details,
                    details: nil
                ))
            }
            break
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/com/godwin/calendar_events/CalendarEventsPlugin.kt
package com.godwin.calendar_events


import android.app.Activity
import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** CalendarEventsPlugin */
class CalendarEventsPlugin : FlutterPlugin, MethodCallHandler, ActivityAware,
    PermissionResultCallback {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var activity: Activity
    private lateinit var result: Result
    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.godwin/calendar_events")
        channel.setMethodCallHandler(this)

        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "requestPermission") {
            this.result = result
            CalendarEventManager.requestPermission(activity)
        } else if (call.method == "getCalendarAccounts") {
            val getResult = CalendarEventManager.getCalendars(context)
            if (getResult is CalendarListSuccess) {
                result.success(getResult.list)
            } else if (getResult is CalendarListFailed) {
                result.error(
                    getResult.error.errorCode,
                    getResult.error.details,
                    null,
                )
            }
        } else if (call.method == "checkPermission") {
            val havePermission = CalendarEventManager.checkCalendarPermission(context)
            if (havePermission) {
                result.success(1)
            } else {
                result.success(0)
            }
        } else if (call.method == "requestSync") {
            try {
                val map = call.arguments as Map<*, *>
                val accountName = map["accountName"] as String
                val accountType = map["accountType"] as String

                CalendarEventManager.requestSync(accountName, accountType)
                result.success(1)
            } catch (e: Exception) {
                e.printStackTrace()
                result.error("invalid_data", e.message, null)
            }
        } else if (call.method == "addEvent") {
            try {
                val map = call.arguments as Map<*, *>
                val event = CalendarUtil.convertToEvent(map)
                val eventResult = CalendarEventManager.addEventToCalendar(context, event)
                if (eventResult is CalendarEventSuccess) {
                    val eventMap = hashMapOf<String, String>()
                    eventMap["eventId"] = eventResult.event.eventId.toString()
                    result.success(eventMap)
                } else if (eventResult is CalendarEventFailed) {
                    result.error(eventResult.error.errorCode, eventResult.error.details, null)
                }
            } catch (e: Exception) {
                e.printStackTrace()
                result.error("invalid_data", e.message, null)
                return
            }
        } else if (call.method == "updateEvent") {
            try {
                val map = call.arguments as Map<*, *>
                val event = CalendarUtil.convertToEvent(map)
                val eventResult = CalendarEventManager.updateEvent(context, event)
                if (eventResult is CalendarEventSuccess) {
                    val eventMap = hashMapOf<String, String>()
                    eventMap["eventId"] = eventResult.event.eventId.toString()
                    result.success(eventMap)
                } else if (eventResult is CalendarEventFailed) {
                    result.error(eventResult.error.errorCode, eventResult.error.details, null)
                }
            } catch (e: Exception) {
                e.printStackTrace()
                result.error("invalid_data", e.message, null)
                return
            }
        } else if (call.method == "deleteEvent") {
            try {
                val map = call.arguments as Map<*, *>
                val eventId = (map["eventId"] as String?)
                val eventResult = CalendarEventManager.deleteEvent(context, eventId)
                if (eventResult is CalendarEventSuccess) {
                    val eventMap = hashMapOf<String, String>()
                    eventMap["eventId"] = eventResult.event.eventId.toString()
                    result.success(eventMap)
                } else if (eventResult is CalendarEventDeleteSuccess) {
                    val eventMap = hashMapOf<String, String>()
                    eventMap["eventId"] = eventResult.eventId
                    result.success(eventMap)
                } else if (eventResult is CalendarEventFailed) {
                    result.error(eventResult.error.errorCode, eventResult.error.details, null)
                }

            } catch (e: Exception) {
                e.printStackTrace()
                result.error("invalid_data", e.message, null)
                return
            }
        } else {
            result.notImplemented()
        }
    }


    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        CalendarEventManager.setActivityBinding(binding)
        CalendarEventManager.setPermissionCallback(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        CalendarEventManager.setActivityBinding(binding)
        CalendarEventManager.setPermissionCallback(this)
    }

    override fun onDetachedFromActivity() {
        CalendarEventManager.setActivityBinding(null)
    }

    override fun onSuccess() {
        result.success(1)
    }

    override fun onFailed(error: CalendarError) {
        result.error(error.errorCode, error.details, null)
    }

}
